app/controllers/admins_controller.rb

```diff
@@ -402,18 +408,28 @@ def runMoss
         
         # Copy their submission over
         FileUtils.cp(subFile,stuDir)
-        require 'rubygems/package'
 
-        # If we need to unarchive this file, then do so
+        # If we need to unarchive this file, then create archive reader
+        arch_type = params["archiveCmd"][ass.id.to_s]
+        arch_path = "#{stuDir}/#{sub.filename}"
+        if arch_type == "tar" then
+          f = File.new(arch_path)
+          archive_extract = Gem::Package::TarReader.new(f)
+          archive_extract.rewind
+        elsif arch_type == "zip" then
+          archive_extract = Zip::File.open(arch_path)
+        elsif arch_type == "tar.gz" then
```

Comment:
Currently, it's the instructor's job to pick an archive type (`plain file`, `tar`, `tgz`, or `zip`), and we extract the archive. Those extensions are only an internal way to represent the type of the archive that the instructor picked, rather than the actual extension.